### PR TITLE
Joins: Use the new struct in various transformations.

### PR DESCRIPTION
**Pull request by wangandi**
Removed `prior_arities` and co from places where I had not done so yet.

Also, this is a new opportunity to rename `JoinInputMapper` if desired.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4505)

### CODE REVIEW

File: src/expr/src/relation/mod.rs [hunk 1]
```diff
@@ -338,23 +339,11 @@ impl RelationExpr {
...
-                    .enumerate()
-                    .flat_map(|(r, a)| std::iter::repeat(r).take(*a))
-                    .collect::<Vec<_>>();
+                // It is important the `new_from_input_types` constructor is
+                // used. Otherwise, Materialize may potentially end up in an
+                // infinite loop.
+                let input_mapper =
+                    join_input_mapper::JoinInputMapper::new_from_input_types(&input_types);
 
                 // A relation's uniqueness constraint holds if there is a
```
Comment (wangandi) [inline src/expr/src/relation/mod.rs:348]: By the way, does this hold anymore given that we no longer reorder join inputs after deciding on an implementation?

Comment (frankmcsherry) [reply src/expr/src/relation/mod.rs:348]: I'm not sure I understand! I think the statement of the comment is true about a join, independent of the implementation or the order of the inputs. It just means to say, if you have A x B x C then if B and C have unique keys that are used appropriately in the join, then the join result has the same keys as in A. I think this reasoning is true starting from A, or B, or C.

**Review by justinj — COMMENTED** [review]
Code LGTM, I don't have a strong feeling on changing the name, this seems fine but I'm also up for something more specific.

> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4505#-:-MJcuIEH1fvJnXGNn_hA:b-gsqnie)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @frankmcsherry, @justinj, and @wangandi)

File: src/expr/src/relation/mod.rs (hunk 1 above)
Comment (wangandi) [reply src/expr/src/relation/mod.rs:348]: I see. I was confused by why it is always the case that the only key potentially retained is from the first input. I guess it's that the uniqueness constraint holds regardless of join input order, but we currently choose to only detect retained keys if it came from the first input.

So if we had:
`create table foo (a primary key, b)`
`create table bar (a primary key, b)`
and a RelationExpr::Join with the inputs in the order `[foo, bar]`:
If the join constraint is `foo.b = bar.a`, then we know that `foo.a` remains unique.
If the join constraint is `foo.a = bar.b`, then `bar.a` will remain unique, but this method does not detect that.

Comment (frankmcsherry) [reply src/expr/src/relation/mod.rs:348]: That's right!

File: src/expr/src/relation/mod.rs [hunk 2]
```diff
@@ -363,24 +352,27 @@ impl RelationExpr {
...
                         }
                     }
+                }
+                let remains_unique = (1..inputs.len()).all(|index| {
                     input_types[index].keys.iter().any(|ks| {
                         ks.iter()
-                            .all(|k| prior_bound.contains(&&(prior_arities[index] + k)))
+                            .all(|k| column_with_prior_bound_by_input[index - 1].contains(k))
                     })
                 });
```
Comment (frankmcsherry) [inline src/expr/src/relation/mod.rs:377]: I'm tempted to say we should move all of this to the `JoinMapper` struct, as it seems to contain re-usable information about type structure. In particular, whatever analysis we want to use about how to find the unique keys of a join. There are other places in the code where we want to know about unique keys (e.g. in redundant join elision) and I suspect that we might want to just lift them out. If `RelationExpr` wrapped structs in each of its variants, we would move the logic to that struct for example.

Comment (frankmcsherry) [reply src/expr/src/relation/mod.rs:377]: I think @justinj might have a dissenting opinion, in that key structure may be one of the things that should be un-baked from the "type".

File: src/expr/src/relation/mod.rs [hunk 3]
```diff
@@ -363,24 +352,27 @@ impl RelationExpr {
                 //
                 // We are going to use the uniqueness constraints of the
                 // first relation, and attempt to use the presented order.
```
Comment (frankmcsherry) [inline src/expr/src/relation/mod.rs:354]: This comment seems less relevant for the code that now follows it. Previously, it explained the structure of the iteration that was just about to follow. Now, it doesn't seem to do that (at least, I don't understand the new code, and this comment is where I would love to see an explanation!).

**Review by frankmcsherry — APPROVED** [review]
This looks mostly good to me. I left some comments around the code that I did not understand (determining the type of joins, whose structure seems to have changed by more than some utility functions).

File: src/expr/src/relation/mod.rs (hunk 2 above)
Comment (justinj) [reply src/expr/src/relation/mod.rs:377]: I've come to just think of what we call a "type" as "logical properties," so I have no strong feelings at this point about it

**Pull request merged**